snuba/datasets/metrics_aggregate_processor.py

```diff
@@ -0,0 +1,135 @@
+
+    @abstractmethod
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    def process_message(
+        self, message: Mapping[str, Any], metadata: KafkaMessageMetadata
+    ) -> Optional[ProcessedMessage]:
+        if not self._should_process(message):
+            return None
+
+        timestamp = _ensure_valid_date(datetime.utcfromtimestamp(message["timestamp"]))
+        assert timestamp is not None
+
+        keys = []
+        values = []
+        tags = message["tags"]
+        assert isinstance(tags, Mapping)
+        for key, value in sorted(tags.items()):
+            assert key.isdigit()
+            keys.append(int(key))
+            assert isinstance(value, int)
+            values.append(value)
+
+        processed = [
+            {
+                "org_id": ClickhouseInt(message["org_id"]),
+                "project_id": ClickhouseInt(message["project_id"]),
+                "metric_id": ClickhouseInt(message["metric_id"]),
+                "timestamp": ClickhouseUnguardedExpression(
+                    f"toStartOfInterval(toDateTime('{timestamp.isoformat()}'), toIntervalSecond({granularity}))"
+                ),
+                "tags.key": ClickhouseNumArray(keys),
+                "tags.value": ClickhouseNumArray(values),
+                **self._process_values(message),
+                "retention_days": ClickhouseInt(message["retention_days"]),
+                "partition": metadata.partition,
+                "offset": metadata.offset,
+                "granularity": ClickhouseInt(granularity),
+            }
+            for granularity in self.GRANULARITIES_SECONDS
+        ]
+        return AggregateInsertBatch(processed, None)
+
+
+class SetsAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return message["type"] is not None and message["type"] == "s"
+
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        values = message["value"]
+        for v in values:
+            assert isinstance(v, int), "Illegal value in set. Int expected: {v}"
+        value_array = "[" + ",".join([str(v) for v in values]) + "]"
+
+        return {
+            "value": ClickhouseUnguardedExpression(
+                f"arrayReduce('uniqState', {value_array})"
+            )
+        }
+
+
+class CounterAggregateProcessor(MetricsAggregateProcessor):
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        return message["type"] is not None and message["type"] == "c"
+
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        value = message["value"]
+        assert isinstance(
+            value, (int, float)
+        ), "Illegal value for counter value. Int/Float expected {value}"
+
+        return {
+            "value": ClickhouseUnguardedExpression(
+                f"arrayReduce('sumState', [{value}])"
```

Comment:
Pretty sure `arrayReduce(sumState, [1,2,3,4,5])` equals `sum([1,2,3,4,5])`. Or just compute the sum directly here. I think it may work.

Reply:
I think it would equal `sumState(15)` but that just inserting 15 directly into the table will cause a datatype error (float64 vs. aggregatefunction). Will confirm locally, but I agree it makes sense to do the sum internally

Reply:
The arrayReduce is actually necessary or I get an error like 

```
INSERT INTO metrics_distributions_local (org_id, project_id, metric_id, granularity, timestamp, retention_days, tags.key, tags.value, percentiles, min, max, avg, sum, count) VALUES


Exception on client:
Code: 184. DB::Exception: Aggregate function minState(24.) is found in wrong place in query
```

for
```
insert into metrics_distributions_local (org_id,project_id,metric_id,granularity,timestamp,retention_days,tags.key,tags.value,percentiles,min,max,avg,sum,count)values (1, 2, 8, 10, toStartOfInterval(toDateTime('2022-01-05T21:35:10'), toIntervalSecond(10)), 90, [4, 5, 6, 9], [117, 7, 91, 134], arrayReduce('quantilesState(0.5,0.75,0.9,0.95,0.99)', [24.0, 80.0, 119.0, 146.0, 182.0]), minState(24.0), maxState(182.0), arrayReduce('avgState', [24.0, 80.0, 119.0, 146.0, 182.0]), sumState(551.0), countState(5))
```

I think the array reduction makes the data types line up for Clickhouse internally (didn't see a lot of good results when Googling this error message).